### PR TITLE
Saved search

### PR DESCRIPTION
**Pull request by bserdar**

### CODE REVIEW

File: crud/src/main/java/com/redhat/lightblue/mediator/Mediator.java [hunk 1]
```diff
@@ -91,8 +91,8 @@
 
     private static final Path OBJECT_TYPE_PATH = new Path("objectType");
 
-    private final Metadata metadata;
-    private final Factory factory;
+    public final Metadata metadata;
+    public final Factory factory;
```
Comment (bvulaj) [inline crud/src/main/java/com/redhat/lightblue/mediator/Mediator.java]: Necessary?

File: misc/src/main/java/com/redhat/lightblue/savedsearch/FindRequestBuilder.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    /**
+     * Builds a find request from the saved search by rewriting the
+     * search bound parameters using parameter values. The search
+     * query components can be either strings, or JSON objects. If
+     * they are strings, the parameters are replaced first, then the
+     * resulting json string is converted to a query/projection. If
+     * the components are JSON objects, then the values of those json
+     * objects are replaced with parameter values.
+     */
+    public static FindRequest buildRequest(JsonNode savedSearch,
```
Comment (bvulaj) [inline misc/src/main/java/com/redhat/lightblue/savedsearch/FindRequestBuilder.java]: What's the workflow for the user here?  Get the saved search from SavedSearchCache and then pass it in here with their params?

File: misc/src/main/java/com/redhat/lightblue/savedsearch/SavedSearchCache.java [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+     *
+     * @param m Mediator instance
+     * @param clid The client id
+     * @param searchName name of the saved search
+     * @param entit Name of the entity
+     * @param version Entity version the search should run on
+     *
+     * @return The saved search document, or null if not found
+     */
+    public JsonNode getSavedSearch(Mediator m,
```
Comment (bvulaj) [inline misc/src/main/java/com/redhat/lightblue/savedsearch/SavedSearchCache.java]: Can we name these a little more intuitively?  `getSavedSearch` and `getSavedSearchFromDb` or something like that.

File: misc/src/main/metadata/lb-saved-search.json [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                        }
+                    }
+                }
+            },
+            "query":{
+                "type":"any",
+                "constraints":{
+                    "required":true
+                },
+                "description":"The query. Bound parameters are given as ${parameter}. This could be a strig, or json "
```
Comment (bvulaj) [inline misc/src/main/metadata/lb-saved-search.json]: typo

File: crud/src/main/java/com/redhat/lightblue/mediator/Mediator.java (hunk 1 above)
Comment (bserdar) [inline crud/src/main/java/com/redhat/lightblue/mediator/Mediator.java]: Used in getSavedSearch. No need to hide these as they're final, either add getters, or make public. This seems more straight forward. Without exposing these, I have to go into configuration to get an instance of Metadata, and that's introducing cyclic dependency,

File: misc/src/main/java/com/redhat/lightblue/savedsearch/FindRequestBuilder.java (hunk 2 above)
Comment (bserdar) [inline misc/src/main/java/com/redhat/lightblue/savedsearch/FindRequestBuilder.java]: The implementation is in rest, I'll push the PR once this is in. The flow is: parse URL, get saved search from the cache, build request using url params, and run the search.

File: misc/src/main/metadata/lb-saved-search.json (hunk 4 above)
Comment (bserdar) [inline misc/src/main/metadata/lb-saved-search.json]: fixed

File: misc/src/main/java/com/redhat/lightblue/savedsearch/FindRequestBuilder.java (hunk 2 above)
Comment (bserdar) [inline misc/src/main/java/com/redhat/lightblue/savedsearch/FindRequestBuilder.java]: Added more documentation describing the workflow.

File: misc/src/main/metadata/lb-saved-search.json [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            "entity": {
+                "type":"string",
+                "constraints":{
+                    "required":true
+                },
+                "description":"The name of the entity search works on"
+            },
+            "versions": {
+                "type":"array",
+                "description": "Entity versions this search can be applied to. It is a partial version match. A 1.0 value in this array means the search will be used for all 1.0.x versions as well as 1.0 version. A null value means the search will be used for all entity versions that are not explicitly selected by other elements in the array.",
```
Comment (paterczm) [inline misc/src/main/metadata/lb-saved-search.json]: When will the default entity version be used?

**Comment by bserdar** [comment]
In the most intuitive way, and after several iterations, I decided that it
is this:

If you call:
     /search/getById/user?id=1
i.e. no version: This uses the default version of user. Whatever that
default is, if there is a saved search for that specific version, that will
be used. Otherwise, saved search containing null version will be used.

If you call:
   /search/getById/user/1.0.0?id=1
This will use the user:1.0.0 version. If there is a search for 1.0.0, 1.0,
or 1, that will be used. Otherwise, search containing null version will be
used.

**Pull request merged**